test/test_autograd.py

```diff
@@ -6207,6 +6205,70 @@ def get_refs(with_backward):
         gc.collect()
         self.assertTrue(ref.expired())
 
+    def test_create_graph_and_full_backward_hook_cycle(self):
+        # If BackwardHook saves grad_output, it can create a cycle when we perform backward
+        # with create_graph=True
+        #
+        #   grad_output -> grad_output.grad_fn -> graph -> hook -> grad_output
+        #
+        class TestCls():
+            # Dummy class for the purpose of creating a weakref
+            pass
+
+        def get_ref(input_requires_grad, nb_hooks, remove_hook_during_backward):
+            t = torch.randn(10, requires_grad=input_requires_grad)
+            a = torch.tensor(1., requires_grad=True)
+
+            def f(*args):
+                # Removing the hooks during backward doesn't matter -
+                # All the hooks will still be executed, and counter will still go to zero
+                if remove_hook_during_backward:
+                    handles[0].remove()
+                return None
+
+            class Test(nn.Module):
+                def forward(self, x):
+                    tmp = a ** 2
+                    # register a hook to try to remove the full backward hook during backward
+                    tmp.grad_fn.register_hook(f)
+                    return tmp * x ** 2
+            mod = Test()
+
+            handles = []
+            for _ in range(nb_hooks):
+                handles.append(mod.register_full_backward_hook(lambda a, b, c: None))
+
+            tmp = mod(t)
+
+            # Save dummy object to graph and get a weak ref to it
+            test = TestCls()
+            ref = weakref.ref(test)
+            tmp.grad_fn.metadata["a"] = test
+
+            with warnings.catch_warnings(record=True) as w:
+                tmp.exp().sum().backward(create_graph=True)
+            self.assertTrue(len(w) <= 1)
+            if len(w) == 1:
```

Comment:
Why would `len(w)` be 0 here?
If it is because of a warn_once, there is a context manager to force these to always warn and make this check more reliable.